### PR TITLE
device: rework how we do struct device extern's

### PR DESCRIPTION
**Pull request by galak**
....
This lets use define DEVICE_ORD_MACRO as needed so if we want to
do different things based on it we can.
....

Signed-off-by: Kumar Gala <galak@kernel.org>

### CODE REVIEW

File: scripts/dts/gen_defines.py [hunk 1]
```diff
@@ -146,34 +146,19 @@ def main():
...
-            print(f"extern const struct device DEVICE_DT_NAME_GET(DT_{node.z_path_id}); /* dts_ord_{node.dep_ordinal} */",
-                  file=dev_header_file)
-
-        print("", file=dev_header_file)
-        print("#ifdef __cplusplus", file=dev_header_file)
-        print("}", file=dev_header_file)
-        print("#endif", file=dev_header_file)
-        print("", file=dev_header_file)
-        print("#endif /* DEVICE_EXTERN_GEN_H */", file=dev_header_file)
+            print(f"DEVICE_ORD_MACRO(DT_{node.z_path_id}); /* dts_ord_{node.dep_ordinal} */",
```
Comment (alevkoy) [inline scripts/dts/gen_defines.py:160]: This produces a non-self-contained header that takes DEVICE_ORD_MACRO as an implicit parameter. I'm familiar with the pattern, but I don't like it very much. It looks weird if you're unfamiliar with the code, and there are some tricky ways to use it incorrectly. The header is internal, so it's not terrible.

However, all other things being equal, I'd prefer it if the script produced a self-contained header, and the client files included it at the top and then called into something that it defined. Something like
```
#include "dts_ord.h"

#define DEVICE_MACRO(ord) \
    extern const struct device DEVICE_DT_NAME_GET(ord)

DT_ORD_FOREACH(DEVICE_MACRO);
```

I think in this scenario, you could have dts_ord.h look like this:
```
// Normal header include guards
#define DT_ORD_FOREACH(fn) \
    fn(DT_N_S_foo); \
    fn(DT_N_S_bar); \
    // ...
```

I think the Python script would be very close to how it is now with just the for loop body changed and an extra line before it.

That's a little bit weird, in that all the ordinal references end up in one giant macro, but the caller doesn't need to know that, and the `#include` and invocation look normal. To my mind, that would be an improvement.

**Review by alevkoy — COMMENTED** [review]
I like what this allows us to do by exposing the device tree contents to C in a reusable way. I'd like it better if it were more self-contained. See my suggestion for how we might do that.

File: include/zephyr/device.h [hunk 2]
```diff
@@ -950,13 +950,16 @@ BUILD_ASSERT(sizeof(device_handle_t) == 2, "fix the linker scripts");
 	Z_INIT_ENTRY_DEFINE(DEVICE_NAME_GET(dev_name), init_fn,		\
 		(&DEVICE_NAME_GET(dev_name)), level, prio)
 
+/* device_extern is generated based on devicetree nodes */
+#undef DEVICE_ORD_MACRO
+#define DEVICE_ORD_MACRO(x) extern const struct device DEVICE_DT_NAME_GET(x)
+
+#include <device_ord.h>
```
Comment (gmarull) [inline include/zephyr/device.h:957]: why do we need undef here? shouldn't `DEVICE_ORD_MACRO` be in `device_ord.h`? I guess it's related to topics discussed in dev-review (sorry I couldn't join), could other usages  be detailed?

**Review by yperess — APPROVED** [review]
I like both this and Abe's approach. Whichever we go with I'm happy, thanks!

**Comment by mbolivar-nordic** [comment]
I have another idea I'll post soon

**Comment by mbolivar-nordic** [comment]
@galak @alevkoy I want to suggest this instead:

 https://github.com/zephyrproject-rtos/zephyr/pull/48092

**Comment by galak** [comment]
Closing my draft version as I like @mbolivar-nordic version of how to handle this better.

**Pull request closed without merging**